cmd/gke-gcloud-auth-plugin/main_test.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientauth "k8s.io/client-go/pkg/apis/clientauthentication/v1beta1"
+)
+
+func TestGcloudPlugin(t *testing.T) {
+	newYears := time.Date(2022, 1, 1, 0, 0, 0, 0, time.UTC)
+
+	tcs := []struct {
+		name   string
+		config string
+
+		wantErr    bool
+		wantStatus *clientauth.ExecCredentialStatus
+	}{
+		{
+			name: "good",
+			config: `
+{
+  "credential": {
+    "access_token": "ya29.t0k3n",
+    "token_expiry": "2022-01-01T00:00:00Z"
+  }
+}
+`,
+			wantStatus: &clientauth.ExecCredentialStatus{
+				Token:               "ya29.t0k3n",
+				ExpirationTimestamp: &metav1.Time{Time: newYears},
+			},
+		},
+		{
+			name: "all good with details",
+			config: `
+{
+  "configuration": {
+    "active_configuration": "default",
+    "properties": {
+      "compute": {
+        "region": "hoth-echobase1",
+        "zone": "hoth-echobase1-c"
+      },
+      "core": {
+        "account": "chewbacca@millenium.falcon",
+        "disable_usage_reporting": "True",
+        "project": "the-resistance"
+      }
+    }
+  },
+  "credential": {
+    "access_token": "ya29.t0k3n",
+    "token_expiry": "2022-01-01T00:00:00Z"
+  }
+}
+`,
+			wantStatus: &clientauth.ExecCredentialStatus{
+				Token:               "ya29.t0k3n",
+				ExpirationTimestamp: &metav1.Time{Time: newYears},
+			},
+		},
+		{
+			name:    "empty response from gcloud",
+			config:  "",
+			wantErr: true,
+		},
+		{
+			name:    "response with empty json object from gcloud",
+			config:  "{}",
+			wantErr: true,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			p := &plugin{
+				readGcloudConfigRaw: func() ([]byte, error) {
+					return []byte(tc.config), nil
+				},
+			}
+
+			creds, err := p.execCredential()
```

Comment:
Nit: Test at p.plugin (the main entry point)?  Or move the JSON serialization into main?

Reply:
Testing against the json makes the tests a lot harder to maintain. Moving to main feels like a loophole. I'd buy adding a unit test for run in it's entirety but I might defer that to a followup up :)

Reply:
Fair enough